File: src/main/scala/jk_5/nailed/server/world/NailedDefaultWorldProviders.scala
/*
 * Nailed, a Minecraft PvP server framework
 * Copyright (C) jk-5 <http://github.com/jk-5/>
 * Copyright (C) Nailed team and contributors <http://github.com/nailed/>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the MIT License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the MIT License along with
 * this program. If not, see <http://opensource.org/licenses/MIT/>.
 */

package jk_5.nailed.server.world

import jk_5.nailed.api.world.{DefaultWorldProviders, Dimension, WorldProvider}

/**
 * No description given
 *
 * @author jk-5
 */
object NailedDefaultWorldProviders extends DefaultWorldProviders {
  private sealed trait IdTracked extends WorldProvider {
    private var id: Int = _
    override final def getId = this.id
    override final def setId(id: Int) = this.id = id
    override def getOptions: String = null
  }
  override def getVoidProvider: WorldProvider = new WorldProvider with IdTracked {
    override def getType = "void"
    override def getDimension = Dimension.OVERWORLD
  }
  override def getOverworldProvider: WorldProvider = new WorldProvider with IdTracked {
    override def getType = "overworld"
    override def getDimension = Dimension.OVERWORLD
  }
  override def getNetherProvider: WorldProvider = new WorldProvider with IdTracked {
    override def getType = "nether"
    override def getDimension = Dimension.NETHER
  }
  override def getEndProvider: WorldProvider = new WorldProvider with IdTracked {
    override def getType = "end"
    override def getDimension = Dimension.END
  }
  override def getFlatProvider(pattern: String): WorldProvider = new WorldProvider with IdTracked {
    override def getType = "flat"
    override def getOptions = pattern
    override def getDimension = Dimension.OVERWORLD
  }
}

File: src/main/java/jk_5/nailed/server/world/NailedDefaultWorldProviders.java
package jk_5.nailed.server.world;

import jk_5.nailed.api.world.DefaultWorldProviders;
import jk_5.nailed.api.world.Dimension;
import jk_5.nailed.api.world.WorldProvider;

import javax.annotation.Nonnull;

public class NailedDefaultWorldProviders implements DefaultWorldProviders {

    private static final NailedDefaultWorldProviders INSTANCE = new NailedDefaultWorldProviders();

    private abstract class IdTracked implements WorldProvider {

        private int id;

        @Override
        public final int getId() {
            return this.id;
        }

        @Override
        public final void setId(int id) {
            this.id = id;
        }

        @Nonnull
        @Override
        public Dimension getDimension() {
            return Dimension.OVERWORLD;
        }

        @Override
        public String getOptions() {
            return null;
        }
    }

    @Nonnull
    @Override
    public WorldProvider getVoidProvider() {
        return new IdTracked() {
            @Nonnull
            @Override
            public String getType() {
                return "void";
            }
        };
    }

    @Nonnull
    @Override
    public WorldProvider getOverworldProvider() {
        return new IdTracked() {
            @Nonnull
            @Override
            public String getType() {
                return "overworld";
            }
        };
    }

    @Nonnull
    @Override
    public WorldProvider getNetherProvider() {
        return new IdTracked() {
            @Nonnull
            @Override
            public String getType() {
                return "nether";
            }

            @Nonnull
            @Override
            public Dimension getDimension() {
                return Dimension.NETHER;
            }
        };
    }

    @Nonnull
    @Override
    public WorldProvider getEndProvider() {
        return new IdTracked() {
            @Nonnull
            @Override
            public String getType() {
                return "end";
            }

            @Nonnull
            @Override
            public Dimension getDimension() {
                return Dimension.END;
            }
        };
    }

    @Nonnull
    @Override
    public WorldProvider getFlatProvider(@Nonnull final String pattern) {
        return new IdTracked() {
            @Nonnull
            @Override
            public String getType() {
                return "flat";
            }

            @Override
            public String getOptions() {
                return pattern;
            }
        };
    }

    public static NailedDefaultWorldProviders instance(){
        return INSTANCE;
    }
}
